### PR TITLE
redid some parts of the binsearch module

### PR DESCRIPTION
**Pull request by SansPapyrus683**
they were awfully inconsistent

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
i'll redo the later parts after this pr is done

**Comment by SansPapyrus683** [comment]
image
screw off bot you're going to mess up the build again

File: content/3_Silver/Binary_Search.mdx [hunk 1]
```diff
@@ -111,25 +111,31 @@ range `[lo,hi]`.
...
-		else hi = mid-1;
-		// if mid does not work, greater values would not work too
-		// so we don't care about them
+		int mid = lo + (hi - lo + 1) / 2;
+		if (test(mid)) {
+			// if mid works, then all numbers smaller than mid also work
+			lo = mid;
+		} else {
+			// if mid does not work, greater values would not work too
+			// so we don't care about them
```
Comment (nwatx) [inline content/3_Silver/Binary_Search.mdx:124]:
```suggestion
			// if mid does not work, greater values would not work either
```

File: content/3_Silver/Binary_Search.mdx [hunk 2]
```diff
@@ -145,23 +151,28 @@ import java.util.function.Predicate;
 
 public class BinarySearch {
 	public static void main(String[] args) {
-		System.out.println(lastTrue(2,10,(x) -> true)); // 10
 		// all numbers satisfy the condition
-		System.out.println(lastTrue(2,10,(x) -> x*x<=30)); // 5
-		System.out.println(lastTrue(2,10,(x) -> false)); // 1
+		System.out.println(lastTrue(2, 10, (x) -> true));  // outputs 10
+		
+		System.out.println(lastTrue(2, 10, (x) -> x * x<=30));  // outputs 5
```
Comment (nwatx) [inline content/3_Silver/Binary_Search.mdx:157]: operator spacing?
```suggestion
		System.out.println(lastTrue(2, 10, (x) -> x * x <= 30));  // outputs 5
```

File: content/3_Silver/Binary_Search.mdx [hunk 3]
```diff
@@ -145,23 +151,28 @@ import java.util.function.Predicate;
...
+			if (test.test(mid)) {
 				// if mid works, then all numbers smaller than mid also work
-			else hi = mid-1;
-			// if mid does not work, greater values would not work too
-			// so we don't care about them
-
+				lo = mid;
+			} else {
+				// if mid does not work, greater values would not work too
+				// so we don't care about them
```
Comment (nwatx) [inline content/3_Silver/Binary_Search.mdx:173]:
```suggestion
				// if mid does not work, greater values would not work either
```

File: content/3_Silver/Binary_Search.mdx [hunk 4]
```diff
@@ -177,33 +188,38 @@ if you're not familiar with the `Predicate` interface used.
...
-			lo = mid+1 # cut the part before mid
-			# notice that we already handled mid itself so we cut it as well
+			lo = mid
 		else:
 			# if mid does not work, greater values would not work too
-			hi = mid-1 # cut mid and after
-	# now res is the answer
-	# if res is l-1 that means no value in the range satisfies the condition
-	return res
+			# so we don't care about them
```
Comment (nwatx) [inline content/3_Silver/Binary_Search.mdx:211]:
```suggestion
			# so we exclude them
```

**Review by nwatx — CHANGES_REQUESTED** [review]

File: content/3_Silver/Binary_Search.mdx [hunk 5]
```diff
@@ -177,33 +186,38 @@ if you're not familiar with the `Predicate` interface used.
...
-print(lastTrue(2, 10, lambda x: True)) # 10
 # all numbers satisfy the condition
-print(lastTrue(2, 10, lambda x : x * x <= 30)) # 5
-print(lastTrue(2, 10, lambda x : False)) # 1
+print(last_true(2, 10, lambda x: True))  # outputs 10
+
+print(last_true(2, 10, lambda x : x * x <= 30))  # outputs 5
+
 # no numbers satisfy the condition
+print(last_true(2, 10, lambda x : False)) #  outputs 1
```
Comment (maggie-j-liu) [inline content/3_Silver/Binary_Search.mdx:220]: be consistent with spacing (pick one of `lambda x: ` and `lambda x : `)

Comment (SansPapyrus683) [reply content/3_Silver/Binary_Search.mdx:220]: oh yeah some of them might slipped my eyes

**Review by maggie-j-liu — COMMENTED** [review]
you didn't format the code from lines 433 to 737, are you going to reformat those too?

**Review by ar492 — COMMENTED** [review]
looks correct

**Comment by SansPapyrus683** [comment]
> i'll redo the later parts after this pr is done

**Review by maggie-j-liu — APPROVED** [review]
i assume this code has been tested?

**Comment by SansPapyrus683** [comment]
dw i test *all* my code
unless it's such a minor change that i am 1000% sure it will work

**Comment by nwatx** [comment]
> dw i test _all_ my code
> unless it's such a minor change that i am 1000% sure it will work

counterexample: modern art 3 :p

**Comment by SansPapyrus683** [comment]
that was a suggestion i made, i don't test my *suggestions*

File: content/3_Silver/Binary_Search.mdx [hunk 6]
```diff
@@ -177,33 +186,38 @@ if you're not familiar with the `Predicate` interface used.
...
 ```python
-def lastTrue(lo, hi, f):
-	""" Binary Search
-	:param f: (lambda function) check a state
-	:param lo: (int) lower bound
-	:param hi: (int) upper bound
-	:return res: (int) the maximum x such that f(x) is true
+from typing import Callable
+
+
```
Comment (ryanchou-dev) [inline content/3_Silver/Binary_Search.mdx:191]:
```suggestion
```

File: content/3_Silver/Binary_Search.mdx [hunk 7]
```diff
@@ -238,16 +258,44 @@ int lastTrue(int lo, int hi, function<bool(int)> f) {
...
 ```
 
 </JavaSection>
 
+<PySection>
+
+```py
+from typing import Callable
+
+
```
Comment (ryanchou-dev) [inline content/3_Silver/Binary_Search.mdx:279]:
```suggestion
```

**Review by TruaShamu — APPROVED** [review]
looks good to me

File: content/3_Silver/Binary_Search.mdx [hunk 8]
```diff
@@ -319,20 +375,23 @@ public class BinarySearch {
 <PySection>
 
 ```python
-def firstTrue(lo, hi, f):
-	# returns smallest x in [lo,hi] that satisfies f
-	# hi+1 if no x satisfies f
-	hi+=1
+from typing import Callable
+
+
```
Comment (ryanchou-dev) [inline content/3_Silver/Binary_Search.mdx:380]:
```suggestion
```

**Review by ryanchou-dev — APPROVED** [review]

File: content/3_Silver/Binary_Search.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply content/3_Silver/Binary_Search.mdx:380]: ryan
it's the convention to have 2 newlines between functions and anything so
i'm just gonna

**Pull request merged**